crates/turborepo-lib/src/process/shared_adaptor.rs

```diff
@@ -0,0 +1,116 @@
+use std::sync::{Arc, Mutex};
+
+use itertools::Either;
+
+use super::{Closed, Open, ProcessManager};
+
+/// A simple adaptor over `ProcessManager` that allows it to be shared between
+/// threads.
+#[derive(Debug)]
+pub struct SharedProcessManager(Arc<Mutex<Either<ProcessManager<Open>, ProcessManager<Closed>>>>);
+
+impl SharedProcessManager {
+    pub fn new() -> Self {
+        Self(Arc::new(Mutex::new(Either::Right(
+            ProcessManager::<Closed>::new(),
+        ))))
+    }
+
+    pub fn start(&self) {
+        let mut lock = self.0.lock().unwrap();
+        // we need ownership of the data in the lock, so replace it with a default value
+        let old = std::mem::replace(&mut *lock, Either::Right(ProcessManager::<Closed>::new()));
+        *lock = old.right_and_then(|v| Either::Left(v.start()));
+    }
+
+    pub fn is_started(&self) -> bool {
+        let lock = self.0.lock().unwrap();
+        lock.is_left()
+    }
+
+    pub async fn wait(&self) {
+        let mut lock = self.0.lock().unwrap();
+        // we need ownership of the data in the lock, so replace it with a default value
+        let old = std::mem::replace(&mut *lock, Either::Right(ProcessManager::<Closed>::new()));
+        *lock = match old {
+            Either::Left(manager) => Either::Right(manager.wait().await),
+            closed => closed,
+        }
+    }
+
+    pub async fn stop(&self) {
+        let mut lock = self.0.lock().unwrap();
+        // we need ownership of the data in the lock, so replace it with a default value
+        let old = std::mem::replace(&mut *lock, Either::Right(ProcessManager::<Closed>::new()));
+        *lock = match old {
+            Either::Left(manager) => Either::Right(manager.stop().await),
+            closed => closed,
+        }
+    }
+
+    pub fn spawn(
+        &self,
+        command: super::child::Command,
+        timeout: std::time::Duration,
+    ) -> Option<super::child::Child> {
+        let mut lock = self.0.lock().unwrap();
+        lock.as_mut().map_left(|l| l.spawn(command, timeout)).left()
```

Comment:
In the case that the process manager is in the process of stopping, then this function will block on all existing processes exiting before it can return `None`. If we are going to wait until closing finishes, I think we need to at least make this await to avoid blocking a thread just waiting for a lock to become free.

Ideally, we would have the same behavior as Go where we recognize that we're shutting down and immediately return `None`.